astropy/io/ascii/ecsv.py

```diff
@@ -170,15 +177,34 @@ def get_cols(self, lines):
         # Transfer attributes from the column descriptor stored in the input
         # header YAML metadata to the new columns to create this table.
         for col in self.cols:
-            for attr in ('description', 'format', 'unit', 'meta'):
+            for attr in ('description', 'format', 'unit', 'meta', 'subtype'):
                 if attr in header_cols[col.name]:
                     setattr(col, attr, header_cols[col.name][attr])
+
             col.dtype = header_cols[col.name]['datatype']
+            if col.dtype not in ECSV_DATATYPES:
+                raise ValueError(f'datatype {col.dtype!r} of column {col.name!r} '
+                                 f'is not in allowed values {ECSV_DATATYPES}')
+
+            # Subtype is written like "int64[2,null]" and we want to split this
+            # out to "int64" and [2, None].
+            subtype = col.subtype
+            if subtype and '[' in subtype:
+                idx = subtype.index('[')
+                col.subtype = subtype[:idx]
+                col.shape = json.loads(subtype[idx:])
+
             # ECSV "string" means numpy dtype.kind == 'U' AKA str in Python 3
-            if col.dtype == 'string':
-                col.dtype = 'str'
-            if col.dtype.startswith('complex'):
-                raise TypeError('ecsv reader does not support complex number types')
+            for attr in ('dtype', 'subtype'):
+                if getattr(col, attr) == 'string':
+                    setattr(col, attr, 'str')
+                if (getattr(col, attr) or '').startswith('complex'):
+                    raise TypeError('ecsv reader does not support complex number types')
```

Comment:
Combine with the error message above? Also, have to decide on type of exception. I think since `ValueError` is better.

Reply:
Fixed (more-or-less) in 98cd119c. This logic was a bit off because `subtype` is actually allowed to be anything, where if it is not recognized then the reader falls through to reading the column as string. This is to allow user-specific subtypes that may not be readable by the astropy parser.